src/values/AbstractValue.js

```diff
@@ -348,18 +305,39 @@ export default class AbstractValue extends Value {
     }
 
     let realm = this.$Realm;
+    if (realm.pathConditions.length === 0) return this;
     let op = this.kind;
     let result = (() => {
       if (op === "&&" || op === "||") {
         let [left, right] = this.args;
-        let refinedLeft = left instanceof AbstractValue ? left.refineWithPathCondition() : left;
+        invariant(left instanceof AbstractValue); // otherwise the factory would have simplified it.
+        let refinedLeft = left.refineWithPathCondition();
         let refinedRight = right instanceof AbstractValue ? right.refineWithPathCondition() : right;
+        // todo: remove this check when there is an alternative way to indicate that an intrinsic object is nullable
+        if (!(refinedLeft instanceof AbstractObjectValue && refinedLeft.isIntrinsic())) {
+          // true && y <=> y
+          // true || y <=> true
+          if (!refinedLeft.mightNotBeTrue()) return op === "&&" ? refinedRight : refinedLeft;
+          // (x == false) && y <=> x
+          // false || y <=> y
+          if (!refinedLeft.mightNotBeFalse()) return op === "||" ? refinedRight : refinedLeft;
+        }
+        if (refinedLeft.getType() === BooleanValue && refinedRight.getType() === BooleanValue) {
+          // (x: boolean) && true <=> x
+          // x || true <=> true
+          if (!refinedRight.mightNotBeTrue()) return op === "&&" ? refinedLeft : realm.intrinsics.true;
+          // (x: boolean) && false <=> false
```

Comment:
Similar to my comment below, what if "x" has side-effect? I do not think it is safe to optimize it away. Or maybe there is a all-arguments-non-side-effect pre-requirement checking somewhere I missed?

Reply:
It's by construction. Difficult to check/enforce, but very fundamental to the nature of abstract values.